FeedStoreChallenge/Core Data/CoreDataFeedStore.swift

```diff
@@ -0,0 +1,121 @@
+//
+//  CoreDataFeedStore.swift
+//  FeedStoreChallenge
+//
+//  Created by Wilmer Barrios on 23/01/21.
+//  Copyright © 2021 Essential Developer. All rights reserved.
+//
+
+import Foundation
+import CoreData
+
+public final class CoreDataFeedStore: FeedStore {
+	
+	typealias StoredFeedData = (feed: [CDFeedImageItem], timestamp: Date)
+	
+	private let container: NSPersistentContainer
+	private let context: NSManagedObjectContext
+	private let modelName = "FeedStoreDataModel"
+	private let storeURL: URL!
+	
+	public init(storeURL: URL) throws {
+		
+		self.storeURL = storeURL
+		
+		guard let modelURL = Bundle(for: CoreDataFeedStore.self).url(forResource: modelName, withExtension:"momd"),
+			  let model = NSManagedObjectModel(contentsOf: modelURL) else {
+			throw NSError()
+		}
+		
+		let description = NSPersistentStoreDescription(url: storeURL)
+		container = NSPersistentContainer(name: modelName, managedObjectModel: model)
+		container.persistentStoreDescriptions = [description]
+		
+		var loadError: Swift.Error?
+		container.loadPersistentStores { loadError = $1 }
+		try loadError.map { throw $0 }
+		
+		context = container.newBackgroundContext()
+	}
+	
+	public func deleteCachedFeed(completion: @escaping DeletionCompletion) {
+		let context = self.context
+		context.perform {
+			let fetchRequest: NSFetchRequest<NSFetchRequestResult> = CDFeedImage.fetchRequest()
+			let batchDeleteRequest = NSBatchDeleteRequest(fetchRequest: fetchRequest)
+			do {
+				try context.execute(batchDeleteRequest)
+				completion(nil)
+			} catch let e {
+				completion(e)
+			}
+		}
+	}
+	
+	public func insert(_ feed: [LocalFeedImage], timestamp: Date, completion: @escaping InsertionCompletion) {
+		let context = self.context
+		context.perform {
+			let feed_data: [CDFeedImageItem] = feed.toCD(context: context)
+			let storeFeed = CDFeedImage(context: context)
+			storeFeed.feed_data = NSOrderedSet(array: feed_data)
+			storeFeed.feed_timestamp = timestamp
+			do {
+				try context.save()
+				completion(nil)
+			} catch let e {
+				completion(e)
+			}
+		}
+	}
+	
+	public func retrieve(completion: @escaping RetrievalCompletion) {
+		let context = self.context
+		context.perform {
+			let fetchRequest: NSFetchRequest<CDFeedImage> = CDFeedImage.fetchRequest()
+			do {
+				let data = try context.fetch(fetchRequest)
+				let mappedData = self.map(data)
+				guard !mappedData.items.isEmpty else { completion(.empty); return }
```

Comment:
At the moment, the `map` method is too complex. It does too many things.

The guard should be **before** calling `map` - so the implementation can be much simpler.

```suggestion
				guard let feed = data.first else { return completion(.empty) }
				let mappedData = self.map(feed)
```

Moreover, the `map` method doesn't require any instance method/property. So it could be a `static func`. This way, you don't need to capture `self` within the closure.